pkg/resource/deploy/step_generator.go

```diff
@@ -336,21 +323,24 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, *re
 
 					// DeleteBeforeCreate implies that we must immediately delete the resource. For correctness,
 					// we must also eagerly delete all resources that depend directly or indirectly on the resource
-					// being replaced.
+					// being replaced and would be replaced by a change to the relevant dependency.
 					//
 					// To do this, we'll utilize the dependency information contained in the snapshot if it is
 					// trustworthy, which is interpreted by the DependencyGraph type.
 					var steps []Step
 					if sg.opts.TrustDependencies {
-						dependents := sg.plan.depGraph.DependingOn(old)
+						toReplace, err := sg.calculateDependentReplacements(old)
+						if err != nil {
+							return nil, result.FromError(err)
+						}
 
 						// Deletions must occur in reverse dependency order, and `deps` is returned in dependency
 						// order, so we iterate in reverse.
-						for i := len(dependents) - 1; i >= 0; i-- {
-							dependentResource := dependents[i]
+						for i := len(toReplace) - 1; i >= 0; i-- {
+							dependentResource := toReplace[i]
 
 							// If we already deleted this resource due to some other DBR, don't do it again.
-							if sg.deletes[urn] {
+							if sg.deletes[dependentResource.URN] {
```

Comment:
What `urn` was the previous code referring to here?

Reply:
The URN of the root of the DBR graph.